packages/web-components/fast-ssr/src/template-renderer/directives.ts

```diff
@@ -0,0 +1,98 @@
+import { RenderInfo } from "@lit-labs/ssr";
+import {
+    ChildrenDirective,
+    Constructable,
+    ExecutionContext,
+    RefDirective,
+    RepeatDirective,
+    SlottedDirective,
+    ViewTemplate,
+} from "@microsoft/fast-element";
+import { TemplateRenderer } from "./template-renderer.js";
+
+/**
+ * Describes an implementation that can render a directive
+ */
+export interface DirectiveRenderer<T extends Constructable> {
+    render(
+        directive: InstanceType<T>,
+        renderInfo: RenderInfo,
+        source: any,
+        renderer: TemplateRenderer,
+        context: ExecutionContext
+    ): IterableIterator<string>;
+    matcher: T;
+}
+
+export const RepeatDirectiveRenderer: DirectiveRenderer<typeof RepeatDirective> = Object.freeze(
+    {
+        matcher: RepeatDirective,
+        *render(
+            directive: InstanceType<typeof RepeatDirective>,
+            renderInfo: RenderInfo,
+            source: any,
+            renderer: TemplateRenderer,
+            context: ExecutionContext
+        ): IterableIterator<string> {
+            const items = directive.itemsBinding(source, context);
+            const template = directive.templateBinding(source, context);
+            const childContext: ExecutionContext = Object.create(context);
+            childContext.parent = source;
+            childContext.parentContext = context;
+
+            if (template instanceof ViewTemplate) {
+                if (directive.options.positioning) {
+                    for (let i = 0, length = items.length; i < length; i++) {
+                        const ctx: ExecutionContext = Object.create(childContext);
+                        // Match fast-element context creation behavior. Perhaps this should be abstracted
```

Comment:
I've been thinking of creating some better APIs around context exactly like you mention. If you want to create an issue for this I can maybe even work on it next week and then we can update this.